src/server/cmd/pachd/main.go

```diff
@@ -924,7 +936,333 @@ func doFullMode(config interface{}) (retErr error) {
 	})
 	go waitForError("S3 Server", errChan, requireNoncriticalServers, func() error {
 		router := s3.Router(s3.NewMasterDriver(), func() (*client.APIClient, error) {
-			return env.GetPachClient(context.Background()), nil
+			return env.GetPachClient(ctx), nil
+		})
+		server := s3.Server(env.Config().S3GatewayPort, router)
+		certPath, keyPath, err := tls.GetCertPaths()
+		if err != nil {
+			log.Warnf("s3gateway TLS disabled: %v", err)
+			return errors.EnsureStack(server.ListenAndServe())
+		}
+		cLoader := tls.NewCertLoader(certPath, keyPath, tls.CertCheckFrequency)
+		// Read TLS cert and key
+		err = cLoader.LoadAndStart()
+		if err != nil {
+			return errors.Wrapf(err, "couldn't load TLS cert for s3gateway: %v", err)
+		}
+		server.TLSConfig = &gotls.Config{GetCertificate: cLoader.GetCertificate}
+		return errors.EnsureStack(server.ListenAndServeTLS(certPath, keyPath))
+	})
+	go waitForError("Prometheus Server", errChan, requireNoncriticalServers, func() error {
+		http.Handle("/metrics", promhttp.Handler())
+		return errors.EnsureStack(http.ListenAndServe(fmt.Sprintf(":%v", env.Config().PrometheusPort), nil))
+	})
+	go func(c chan os.Signal) {
+		<-c
+		log.Println("terminating; waiting for enterprise server to gracefully stop")
+		var g, _ = errgroup.WithContext(ctx)
+		g.Go(func() error { externalServer.Server.GracefulStop(); return nil })
+		g.Go(func() error { internalServer.Server.GracefulStop(); return nil })
+		g.Wait()
+		log.Println("gRPC server gracefully stopped")
+	}(interruptChan)
+	return <-errChan
+}
+
+func doPausedMode(config interface{}) (retErr error) {
```

Comment:
What you have here makes sense to me, but it would be easier to catch issues if we didn't need to duplicate so much code between modes. I think it could be nice to refactor this file a bit in a separate PR. That being said it's up to you, if you don't want to handle that right now.

Reply:
I agree, but I think that should be a separate PR. Also, the tricky thing is that they are almost but not quite the same, and I worry that a `doAllModes(mode, config)` might end up being its own kind of spaghetti. I think that there is probably an elegant solution, but it requires some thought & planning.